irace/parameters.txt
## Template for parameter description file for Iterated Race.
##
## The format is one parameter per line. Each line contains:
##
## 1: Name of the parameter. An unquoted alphanumeric string,
##    example: ants

## 2: Switch to pass the parameter. A quoted (possibly empty) string,
##    if the value and the switch must be separated, add a space at
##    the end of the string. Example : "--version1 --ants "

## 3: Type. An unquoted single letter, among
##     i: Integer, c: categorical, o: ordinal, r: real.

## 4: For c and o: All possible values, that is, a variable number of
##    quoted or unquoted strings separated by commas within
##    parenthesis. Empty strings and strings containing commas or
##    spaces must be quoted.
##    For i,r: a pair of unquoted numbers representing minimum and
##    maximum values.

## 5: A conditional parameter can be defined according to the values of
##    one or several other parameters.  This is done by adding a
##    character '|' followed by an R expression involving the names of
##    other parameters. This expression must return TRUE if the
##    condition is satisfied, FALSE otherwise.

# 1:          2:                               3:   4:            5:
popsize	      "PopulationSize="	                i   (100, 300)
minspecies    "MinSpecies="                     i   (2, 6)
maxspecies    "MaxSpecies="                     i   (8, 20)
clones        "AllowClones="                    c   (true, false)
youngthresh   "YoungAgeThreshold="              i   (3, 10)       | youngboost > 1.0
youngboost    "YoungAgeFitnessBoost="           r   (1.00, 1.50)
oldthresh     "OldAgeThreshold="                i   (25, 70)      | oldpenalty < 1.0
oldpenalty    "OldAgePenalty="                  r   (0.50, 1.00)
survival      "SurvivalRate="                   r   (0.25, 0.75)
crossover     "CrossoverRate="                  r   (0.50, 0.90)
# The "interspecies" parameter is represented as a percentage, which is later
# converted to a decimal by the targetRunner, i.e., (0.01, 0.50) -> (0.0001, 0.0050)
interspecies  "InterspeciesCrossoverRate="      r   (0.01, 0.50)
mutation      "OverallMutationRate="            r   (0.15, 0.80)
elitism       "Elitism="                        r   (0.00, 0.15)
multipoint    "MultipointCrossoverRate="        r   (0.00, 1.00)
addneuron     "MutateAddNeuronProb="            r   (0.01, 0.15)
remneuron     "MutateRemSimpleNeuronProb="      r   (0.01, 0.15)
addlink       "MutateAddLinkProb="              r   (0.01, 0.15)
remlink       "MutateRemLinkProb="              r   (0.01, 0.15)
# <Weight mutations>
mutweights    "MutateWeightsProb="              r   (0.60, 0.95)
mutsevere     "MutateWeightsSevereProb="        r   (0.10, 0.40)
mutgene       "WeightMutationRate="             r   (0.70, 1.00)
replweight    "WeightReplacementRate="          r   (0.00, 0.40)
maxweight     "MaxWeight="                      i   (6, 20)
# Fictional parameter (WeightMutationMaxProportion): defines the maximum weight mutation
# proportionally to the weight maximum (MaxWeight)
mutweightmax  "WeightMutationMaxProportion="    r   (0.05, 0.30)
# </Weight mutations>
mutbias       "MutateNeuronBiasesProb="         r   (0.00, 0.10)
mutactiv      "MutateNeuronActivationTypeProb=" r   (0.00, 0.01)  | activation > 4
# Each integer in "activation" corresponds to a predefined combination of activation functions
activation    "Activation="                     i   (0, 14)
compat        "CompatThreshold="                i   (2, 10)

# These parameters are related to stagnation, which would require long execution times to be properly tuned
# maxstag       "SpeciesMaxStagnation="           i   (30, 70)
# coevolstag    "DetectCompetetiveCoevolutionStagnation="   c   (true, false)
# killgens      "KillWorstSpeciesEach="           i   (15, 50)      | coevolstag == "true"
# killage       "KillWorstAge="                   i   (10, 30)      | coevolstag == "true"
# deltacoding   "DeltaCoding="                    c   (true, false)
# phased        "PhasedSearching="                c   (true, false)
# simplmpc      "SimplifyingPhaseMPCThreshold="    i   (10, 50)
# simplstag     "SimplifyingPhaseStagnationThreshold=" i (20, 50)
# complstag     "ComplexityFloorGenerations="     i   (25, 60)


irace/initial_config.txt
## Template for specifying initial parameter configurations, for
## example, the default configuration, in irace.
##
## This filename must be specified via the --candidates-file command-line option
## (or candidatesFile in scenario.txt).
##
## The format is one parameter configuration per line, and one parameter per
## column. The first line gives the parameter name corresponding to
## each column (names must match those given in the parameters
## file). Each configuration must satisfy the
## parameter conditions (NA should be used for those parameters that
## are not enabled for a given configuration) and, if given, the
## constraints that describe forbidden configurations.

popsize addneuron remneuron addlink remlink mutbias mutactiv  elitism crossover activation
100     0.05      0.03      0.1     0.05    0.03    0.02      0.1     0.6       1


src/py/params.py
import MultiNEAT as neat


def _default_params():
    params = neat.Parameters()
    params.PopulationSize = 150
    params.DynamicCompatibility = True
    params.AllowClones = True
    params.CompatThreshold = 5.0
    params.CompatThresholdModifier = 0.3
    params.YoungAgeThreshold = 15
    params.SpeciesMaxStagnation = 100
    params.OldAgeThreshold = 35
    params.MinSpecies = 2
    params.MaxSpecies = 10
    params.RouletteWheelSelection = True
    params.RecurrentProb = 0.0
    params.OverallMutationRate = 0.25
    params.MutateWeightsProb = 0.90
    params.WeightMutationMaxPower = 1.0
    params.WeightReplacementMaxPower = 5.0
    params.MutateWeightsSevereProb = 0.01
    params.WeightMutationRate = 0.75
    params.MaxWeight = 20
    params.MutateAddNeuronProb = 0.05
    params.MutateAddLinkProb = 0.1
    params.MutateRemLinkProb = 0.05
    params.MutateRemSimpleNeuronProb = 0.05
    params.EliteFraction = 0.15
    params.CrossoverRate = 0.5

    params.ActivationFunction_UnsignedGauss_Prob = 0.25
    params.ActivationFunction_UnsignedSigmoid_Prob = 0.25
    params.ActivationFunction_UnsignedSine_Prob = 0.25
    params.ActivationFunction_Relu_Prob = 0.25

    params.MutateNeuronTraitsProb = 0
    params.MutateLinkTraitsProb = 0

    return params


def get_params(params_file):
    if params_file is None:
        return _default_params()
    else:
        params = neat.Parameters()
        params.Load(params_file)
        return params


class ParametersWrapper:
    def __init__(self, neat_params):
        self.PopulationSize = neat_params.PopulationSize
        self.DynamicCompatibility = neat_params.DynamicCompatibility
        self.MinSpecies = neat_params.MinSpecies
        self.MaxSpecies = neat_params.MaxSpecies
        self.InnovationsForever = neat_params.InnovationsForever
        self.AllowClones = neat_params.AllowClones
        self.ArchiveEnforcement = neat_params.ArchiveEnforcement
        self.NormalizeGenomeSize = neat_params.NormalizeGenomeSize
        self.CustomConstraints = neat_params.CustomConstraints
        self.YoungAgeThreshold = neat_params.YoungAgeThreshold
        self.YoungAgeFitnessBoost = neat_params.YoungAgeFitnessBoost
        self.SpeciesDropoffAge = neat_params.SpeciesDropoffAge
        self.StagnationDelta = neat_params.StagnationDelta
        self.OldAgeThreshold = neat_params.OldAgeThreshold
        self.OldAgePenalty = neat_params.OldAgePenalty
        self.DetectCompetetiveCoevolutionStagnation = neat_params.DetectCompetetiveCoevolutionStagnation
        self.KillWorstSpeciesEach = neat_params.KillWorstSpeciesEach
        self.KillWorstAge = neat_params.KillWorstAge
        self.SurvivalRate = neat_params.SurvivalRate
        self.CrossoverRate = neat_params.CrossoverRate
        self.OverallMutationRate = neat_params.OverallMutationRate
        self.InterspeciesCrossoverRate = neat_params.InterspeciesCrossoverRate
        self.MultipointCrossoverRate = neat_params.MultipointCrossoverRate
        self.RouletteWheelSelection = neat_params.RouletteWheelSelection
        self.PhasedSearching = neat_params.PhasedSearching
        self.DeltaCoding = neat_params.DeltaCoding
        self.SimplifyingPhaseMPCThreshold = neat_params.SimplifyingPhaseMPCThreshold
        self.SimplifyingPhaseStagnationThreshold = neat_params.SimplifyingPhaseStagnationThreshold
        self.ComplexityFloorGenerations = neat_params.ComplexityFloorGenerations
        self.NoveltySearch_K = neat_params.NoveltySearch_K
        self.NoveltySearch_P_min = neat_params.NoveltySearch_P_min
        self.NoveltySearch_Dynamic_Pmin = neat_params.NoveltySearch_Dynamic_Pmin
        self.NoveltySearch_No_Archiving_Stagnation_Threshold = neat_params.NoveltySearch_No_Archiving_Stagnation_Threshold
        self.NoveltySearch_Pmin_lowering_multiplier = neat_params.NoveltySearch_Pmin_lowering_multiplier
        self.NoveltySearch_Pmin_min = neat_params.NoveltySearch_Pmin_min
        self.NoveltySearch_Quick_Archiving_Min_Evaluations = neat_params.NoveltySearch_Quick_Archiving_Min_Evaluations
        self.NoveltySearch_Pmin_raising_multiplier = neat_params.NoveltySearch_Pmin_raising_multiplier
        self.NoveltySearch_Recompute_Sparseness_Each = neat_params.NoveltySearch_Recompute_Sparseness_Each
        self.MutateAddNeuronProb = neat_params.MutateAddNeuronProb
        self.SplitRecurrent = neat_params.SplitRecurrent
        self.SplitLoopedRecurrent = neat_params.SplitLoopedRecurrent
        self.MutateAddLinkProb = neat_params.MutateAddLinkProb
        self.MutateAddLinkFromBiasProb = neat_params.MutateAddLinkFromBiasProb
        self.MutateRemLinkProb = neat_params.MutateRemLinkProb
        self.MutateRemSimpleNeuronProb = neat_params.MutateRemSimpleNeuronProb
        self.LinkTries = neat_params.LinkTries
        self.RecurrentProb = neat_params.RecurrentProb
        self.RecurrentLoopProb = neat_params.RecurrentLoopProb
        self.MutateWeightsProb = neat_params.MutateWeightsProb
        self.MutateWeightsSevereProb = neat_params.MutateWeightsSevereProb
        self.WeightMutationRate = neat_params.WeightMutationRate
        self.WeightMutationMaxPower = neat_params.WeightMutationMaxPower
        self.WeightReplacementRate = neat_params.WeightReplacementRate
        self.WeightReplacementMaxPower = neat_params.WeightReplacementMaxPower
        self.MaxWeight = neat_params.MaxWeight
        self.MutateActivationAProb = neat_params.MutateActivationAProb
        self.MutateActivationBProb = neat_params.MutateActivationBProb
        self.ActivationAMutationMaxPower = neat_params.ActivationAMutationMaxPower
        self.ActivationBMutationMaxPower = neat_params.ActivationBMutationMaxPower
        self.MinActivationA = neat_params.MinActivationA
        self.MaxActivationA = neat_params.MaxActivationA
        self.MinActivationB = neat_params.MinActivationB
        self.MaxActivationB = neat_params.MaxActivationB
        self.TimeConstantMutationMaxPower = neat_params.TimeConstantMutationMaxPower
        self.BiasMutationMaxPower = neat_params.BiasMutationMaxPower
        self.MutateNeuronTimeConstantsProb = neat_params.MutateNeuronTimeConstantsProb
        self.MutateNeuronBiasesProb = neat_params.MutateNeuronBiasesProb
        self.MinNeuronTimeConstant = neat_params.MinNeuronTimeConstant
        self.MaxNeuronTimeConstant = neat_params.MaxNeuronTimeConstant
        self.MinNeuronBias = neat_params.MinNeuronBias
        self.MaxNeuronBias = neat_params.MaxNeuronBias
        self.MutateNeuronActivationTypeProb = neat_params.MutateNeuronActivationTypeProb
        self.ActivationFunction_SignedSigmoid_Prob = neat_params.ActivationFunction_SignedSigmoid_Prob
        self.ActivationFunction_UnsignedSigmoid_Prob = neat_params.ActivationFunction_UnsignedSigmoid_Prob
        self.ActivationFunction_Tanh_Prob = neat_params.ActivationFunction_Tanh_Prob
        self.ActivationFunction_TanhCubic_Prob = neat_params.ActivationFunction_TanhCubic_Prob
        self.ActivationFunction_SignedStep_Prob = neat_params.ActivationFunction_SignedStep_Prob
        self.ActivationFunction_UnsignedStep_Prob = neat_params.ActivationFunction_UnsignedStep_Prob
        self.ActivationFunction_SignedGauss_Prob = neat_params.ActivationFunction_SignedGauss_Prob
        self.ActivationFunction_UnsignedGauss_Prob = neat_params.ActivationFunction_UnsignedGauss_Prob
        self.ActivationFunction_Abs_Prob = neat_params.ActivationFunction_Abs_Prob
        self.ActivationFunction_SignedSine_Prob = neat_params.ActivationFunction_SignedSine_Prob
        self.ActivationFunction_UnsignedSine_Prob = neat_params.ActivationFunction_UnsignedSine_Prob
        self.ActivationFunction_Linear_Prob = neat_params.ActivationFunction_Linear_Prob
        self.ActivationFunction_Relu_Prob = neat_params.ActivationFunction_Relu_Prob
        self.ActivationFunction_Softplus_Prob = neat_params.ActivationFunction_Softplus_Prob
        self.DontUseBiasNeuron = neat_params.DontUseBiasNeuron
        self.AllowLoops = neat_params.AllowLoops
        self.MutateNeuronTraitsProb = neat_params.MutateNeuronTraitsProb
        self.MutateLinkTraitsProb = neat_params.MutateLinkTraitsProb
        self.DisjointCoeff = neat_params.DisjointCoeff
        self.ExcessCoeff = neat_params.ExcessCoeff
        self.WeightDiffCoeff = neat_params.WeightDiffCoeff
        self.ActivationADiffCoeff = neat_params.ActivationADiffCoeff
        self.ActivationBDiffCoeff = neat_params.ActivationBDiffCoeff
        self.TimeConstantDiffCoeff = neat_params.TimeConstantDiffCoeff
        self.BiasDiffCoeff = neat_params.BiasDiffCoeff
        self.ActivationFunctionDiffCoeff = neat_params.ActivationFunctionDiffCoeff
        self.CompatThreshold = neat_params.CompatThreshold
        self.MinCompatThreshold = neat_params.MinCompatThreshold
        self.CompatThresholdModifier = neat_params.CompatThresholdModifier
        self.CompatThreshChangeInterval_Generations = neat_params.CompatThreshChangeInterval_Generations
        self.CompatThreshChangeInterval_Evaluations = neat_params.CompatThreshChangeInterval_Evaluations

        self.DivisionThreshold = neat_params.DivisionThreshold
        self.VarianceThreshold = neat_params.VarianceThreshold
        self.BandThreshold = neat_params.BandThreshold
        self.InitialDepth = neat_params.InitialDepth
        self.MaxDepth = neat_params.MaxDepth
        self.CPPN_Bias = neat_params.CPPN_Bias
        self.Width = neat_params.Width
        self.Height = neat_params.Height
        self.Qtree_Y = neat_params.Qtree_Y
        self.Qtree_X = neat_params.Qtree_X
        self.Leo = neat_params.Leo
        self.LeoThreshold = neat_params.LeoThreshold
        self.LeoSeed = neat_params.LeoSeed

        self.GeometrySeed = neat_params.GeometrySeed
        self.TournamentSize = neat_params.TournamentSize
        self.EliteFraction = neat_params.EliteFraction
